src/analysis/vector/qgsgeometrysnapper.cpp

```diff
@@ -184,124 +186,43 @@ class Raytracer
-  }
-}
-
-const QgsSnapIndex::Cell *QgsSnapIndex::GridRow::getCell( int col ) const
-{
-  if ( col < mColStartIdx || col >= mColStartIdx + mCells.size() )
-  {
-    return nullptr;
-  }
-  else
-  {
-    return &mCells[col - mColStartIdx];
-  }
-}
-
-QList<QgsSnapIndex::SnapItem *> QgsSnapIndex::GridRow::getSnapItems( int colStart, int colEnd ) const
-{
-  colStart = std::max( colStart, mColStartIdx );
-  colEnd = std::min( colEnd, mColStartIdx + mCells.size() - 1 );
-
-  QList<SnapItem *> items;
-
-  for ( int col = colStart; col <= colEnd; ++col )
-  {
-    items.append( mCells[col - mColStartIdx] );
-  }
-  return items;
-}
-
-///////////////////////////////////////////////////////////////////////////////
-
 QgsSnapIndex::QgsSnapIndex( const QgsPoint &origin, double cellSize )
   : mOrigin( origin )
   , mCellSize( cellSize )
-  , mRowsStartIdx( 0 )
 {
+  mSTRTree = GEOSSTRtree_create_r( QgsGeos::getGEOSHandler(), ( size_t )10 );
 }
 
 QgsSnapIndex::~QgsSnapIndex()
 {
   qDeleteAll( mCoordIdxs );
-}
+  qDeleteAll( mSnapItems );
 
-
-const QgsSnapIndex::Cell *QgsSnapIndex::getCell( int col, int row ) const
-{
-  if ( row < mRowsStartIdx || row >= mRowsStartIdx + mGridRows.size() )
-  {
-    return nullptr;
-  }
-  else
-  {
-    return mGridRows[row - mRowsStartIdx].getCell( col );
-  }
-}
-
-QgsSnapIndex::Cell &QgsSnapIndex::getCreateCell( int col, int row )
-{
-  if ( row < mRowsStartIdx )
-  {
-    for ( int i = row; i < mRowsStartIdx; ++i )
-    {
-      mGridRows.prepend( GridRow() );
-    }
-    mRowsStartIdx = row;
-    return mGridRows.front().getCreateCell( col );
-  }
-  else if ( row >= mRowsStartIdx + mGridRows.size() )
-  {
-    for ( int i = mRowsStartIdx + mGridRows.size(); i <= row; ++i )
-    {
-      mGridRows.append( GridRow() );
-    }
-    return mGridRows.back().getCreateCell( col );
-  }
-  else
-  {
-    return mGridRows[row - mRowsStartIdx].getCreateCell( col );
-  }
+  GEOSSTRtree_destroy_r( QgsGeos::getGEOSHandler(), mSTRTree );
 }
 
 void QgsSnapIndex::addPoint( const CoordIdx *idx, bool isEndPoint )
 {
   QgsPoint p = idx->point();
   int col = std::floor( ( p.x() - mOrigin.x() ) / mCellSize );
   int row = std::floor( ( p.y() - mOrigin.y() ) / mCellSize );
-  getCreateCell( col, row ).append( new PointSnapItem( idx, isEndPoint ) );
+
+  GEOSContextHandle_t geosctxt = QgsGeos::getGEOSHandler();
+#if GEOS_VERSION_MAJOR>3 || GEOS_VERSION_MINOR>=8
+  geos::unique_ptr point( GEOSGeom_createPointFromXY_r( geosctxt, row, col ) );
+#else
+  GEOSCoordSequence *seq = GEOSCoordSeq_create_r( geosctxt, 1, 2 );
+  GEOSCoordSeq_setX_r( geosctxt, seq, 0, row );
+  GEOSCoordSeq_setY_r( geosctxt, seq, 0, col );
+  geos::unique_ptr point( GEOSGeom_createPoint_r( geosctxt, seq ) );
+#endif
+
+  PointSnapItem *item = new PointSnapItem( idx, isEndPoint );
+  GEOSSTRtree_insert_r( geosctxt, mSTRTree, point.get(), item );
+#if GEOS_VERSION_MAJOR>3 || GEOS_VERSION_MINOR<9
+  mSTRTreeItems.push_back( std::move( point ) );
```

Comment:
```suggestion
  mSTRTreeItems.emplace_back( std::move( point ) );
```

(because c++ :shrug: )